File: packages/nodx-js/src/navigation.mjs
import { plainNodeText } from "./ast.mjs";

export function resolveNavigation(doc) {
  const ids = new Map();
  collectIdPaths(doc.body, "", ids);
  const navigations = [];
  collectTocs(doc.body, "", doc, ids, navigations);
  return { navigations };
}

function collectIdPaths(nodes, prefix, ids) {
  nodes.forEach((item, i) => {
    const path = childPath(prefix, i);
    if (item.id !== null) ids.set(item.id, path);
    collectIdPaths(item.children, path, ids);
  });
}

function collectTocs(nodes, prefix, doc, ids, out) {
  nodes.forEach((item, i) => {
    const path = childPath(prefix, i);
    if (item.type === "toc") out.push(resolveToc(item, path, doc, ids));
    collectTocs(item.children, path, doc, ids, out);
  });
}

function resolveToc(toc, path, doc, ids) {
  const role = toc.attrs.role ?? "primary";
  const label = toc.attrs.title ?? defaultNavigationLabel(role);
  const scope = toc.attrs.scope ?? null;
  const sourcePath = scope?.startsWith("#") ? ids.get(scope.slice(1)) : null;
  const scopedNode = sourcePath ? nodeAtPath(doc.body, sourcePath) : null;
  const sourceNodes = scopedNode ? [scopedNode] : doc.body;
  const minLevel = parseLevel(toc.attrs["min-level"]);
  const maxLevel = parseLevel(toc.attrs["max-level"]);
  const depth = parseLevel(toc.attrs.depth) ?? 6;
  const baseLevel = firstHeadingLevel(sourceNodes) ?? 1;
  const effectiveMin = minLevel ?? baseLevel;
  const effectiveMax = maxLevel ?? Math.min(baseLevel + depth - 1, 6);
  const entries = [];
  collectEntries(sourceNodes, "", effectiveMin, effectiveMax, entries);
  return { entries, label, role, scope, tocId: toc.id, tocPath: path };
}

function collectEntries(nodes, prefix, minLevel, maxLevel, out) {
  nodes.forEach((item, i) => {
    const path = childPath(prefix, i);
    if (item.type === "heading" && item.id !== null) {
      const level = headingLevel(item);
      if (level !== null && level >= minLevel && level <= maxLevel) {
        out.push({ id: item.id, level, path, title: plainNodeText(item) });
      }
    }
    collectEntries(item.children, path, minLevel, maxLevel, out);
  });
}

function firstHeadingLevel(nodes) {
  for (const item of nodes) {
    if (item.type === "heading") {
      const level = headingLevel(item);
      if (level !== null) return level;
    }
    const level = firstHeadingLevel(item.children);
    if (level !== null) return level;
  }
  return null;
}

function nodeAtPath(nodes, path) {
  let current = nodes;
  let item = null;
  for (const part of path.split(".")) {
    const index = Number.parseInt(part, 10);
    item = current[index];
    if (!item) return null;
    current = item.children;
  }
  return item;
}

function headingLevel(item) {
  return parseLevel(item.attrs.level);
}

function parseLevel(value) {
  if (typeof value !== "string" || !/^[0-9]+$/.test(value)) return null;
  const level = Number.parseInt(value, 10);
  return level >= 1 && level <= 6 ? level : null;
}

function defaultNavigationLabel(role) {
  if (role === "local") return "In this section";
  if (role === "secondary") return "Secondary navigation";
  if (role === "breadcrumb") return "Breadcrumb";
  return "Table of contents";
}

function childPath(prefix, index) {
  return prefix === "" ? String(index) : prefix + "." + index;
}

File: packages/nodx-py/src/nodx/navigation.py
from .ast import plain_inlines, plain_node_text


def resolve_navigation(doc):
    ids = {}
    collect_id_paths(doc["body"], "", ids)
    navigations = []
    collect_tocs(doc["body"], "", doc, ids, navigations)
    return {"navigations": navigations}


def collect_id_paths(nodes, prefix, ids):
    for i, item in enumerate(nodes):
        path = child_path(prefix, i)
        if item["id"] is not None:
            ids[item["id"]] = path
        collect_id_paths(item["children"], path, ids)


def collect_tocs(nodes, prefix, doc, ids, out):
    for i, item in enumerate(nodes):
        path = child_path(prefix, i)
        if item["type"] == "toc":
            out.append(resolve_toc(item, path, doc, ids))
        collect_tocs(item["children"], path, doc, ids, out)


def resolve_toc(toc, path, doc, ids):
    role = toc["attrs"].get("role", "primary")
    label = toc["attrs"].get("title", default_navigation_label(role))
    scope = toc["attrs"].get("scope")
    if toc["attrs"].get("mode") == "manual":
        entries = []
        collect_manual_entries(toc["children"], ids, entries)
        return {"entries": entries, "label": label, "role": role, "scope": scope, "tocId": toc["id"], "tocPath": path}
    source_path = ids.get(scope[1:]) if isinstance(scope, str) and scope.startswith("#") else None
    scoped_node = node_at_path(doc["body"], source_path) if source_path else None
    source_nodes = [scoped_node] if scoped_node else doc["body"]
    min_level = parse_level(toc["attrs"].get("min-level"))
    max_level = parse_level(toc["attrs"].get("max-level"))
    depth = parse_level(toc["attrs"].get("depth")) or 6
    base_level = first_heading_level(source_nodes) or 1
    effective_min = min_level or base_level
    effective_max = max_level or min(base_level + depth - 1, 6)
    entries = []
    collect_entries(source_nodes, "", effective_min, effective_max, entries)
    return {"entries": entries, "label": label, "role": role, "scope": scope, "tocId": toc["id"], "tocPath": path}


def collect_entries(nodes, prefix, min_level, max_level, out):
    for i, item in enumerate(nodes):
        path = child_path(prefix, i)
        if item["type"] == "heading" and item["id"] is not None:
            level = heading_level(item)
            if level is not None and min_level <= level <= max_level and item["attrs"].get("toc-hidden") != "true":
                out.append({
                    "id": item["id"],
                    "level": parse_level(item["attrs"].get("toc-level")) or level,
                    "path": path,
                    "title": item["attrs"].get("toc", plain_node_text(item)),
                })
        collect_entries(item["children"], path, min_level, max_level, out)


def collect_manual_entries(nodes, ids, out):
    for item in nodes:
        collect_manual_inline_entries(item["inlines"], ids, out)
        collect_manual_entries(item["children"], ids, out)


def collect_manual_inline_entries(inlines, ids, out):
    for item in inlines:
        if item["type"] == "link":
            if item["target"].startswith("#"):
                id_ = item["target"][1:]
                out.append({"id": id_, "level": 1, "path": ids.get(id_, ""), "title": plain_inlines(item["label"])})
            collect_manual_inline_entries(item["label"], ids, out)
        elif item["type"] in ("strong", "em", "mark", "sub", "sup", "span"):
            collect_manual_inline_entries(item["children"], ids, out)


def first_heading_level(nodes):
    for item in nodes:
        if item["type"] == "heading":
            level = heading_level(item)
            if level is not None:
                return level
        level = first_heading_level(item["children"])
        if level is not None:
            return level
    return None


def node_at_path(nodes, path):
    current = nodes
    item = None
    for part in path.split("."):
        index = int(part)
        if index >= len(current):
            return None
        item = current[index]
        current = item["children"]
    return item


def heading_level(item):
    return parse_level(item["attrs"].get("level"))


def parse_level(value):
    if not isinstance(value, str) or not value.isdigit():
        return None
    level = int(value)
    return level if 1 <= level <= 6 else None


def default_navigation_label(role):
    if role == "local":
        return "In this section"
    if role == "secondary":
        return "Secondary navigation"
    if role == "breadcrumb":
        return "Breadcrumb"
    return "Table of contents"


def child_path(prefix, index):
    return str(index) if prefix == "" else prefix + "." + str(index)
